django/db/models/fields/related.py

```diff
@@ -1030,8 +1049,44 @@ def _check_on_delete(self):
                     id="fields.E321",
                 )
             ]
-        else:
-            return []
+        elif (
+            on_delete in [DB_CASCADE, DB_SET_NULL, DB_RESTRICT]
+            and hasattr(self.model, "_meta")
+            and (
+                any(  # generic relation
+                    hasattr(field, "bulk_related_objects")
+                    for field in self.model._meta.private_fields
+                )
+                or any(  # generic foreign key
+                    hasattr(field, "get_content_type")
+                    for field in self.model._meta.private_fields
+                )
+            )
+        ):
+            return [
+                checks.Error(
+                    f"Field specifies unsupported on_delete={on_delete} on model "
+                    "declaring a GenericForeignKey.",
+                    hint="Change the on_delete rule to a non DB_* method",
+                    obj=self,
+                    id="fields.E345",
+                )
+            ]
+        elif related_model_status := self._has_related_models_with_db_cascading(
+            self.model, on_delete
+        ):
+            return [
+                checks.Error(
+                    "Using normal cascading with DB cascading referenced model is "
+                    "prohibited "
+                    f"Related model is {related_model_status.get('model')} "
+                    f"Related field is {related_model_status.get('field')}",
```

Comment:
Can we do this for the `GenericForeignKey` check above?

Reply:
The field object is already being passed in the Error message so I don't think it would be necessary as the field name is already being shown to the user.

Reply:
I thought there was the field we're checking (`self`) and a different field was the GFK and only `self` was being shown in the error message. If both are seen, that's good.

Reply:
That would be a nice to have ig as we are not supporting db cascade to any kind of GFK.